File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        if (n == 0) return {-1, -1};
        int first = findfirst(nums, 0, n - 1, target);
        if (first == -1) return {-1, -1};
        int last = findlast(nums,0,n-1,target);
        return {first, last};
    }
private:
    int findfirst(vector<int>& nums, int low, int high, int target){
            if(low>high) return -1;
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(mid==0 || nums[mid-1]<target){
                    return mid;
                } else return findfirst(nums,low,mid-1,target);
            } 
            else if(target > nums[mid]){
                return findfirst(nums,mid+1,high,target);
            }
            return findfirst(nums,low,mid-1,target);
        }
    int findlast(vector<int>& nums, int low, int high, int target){
            if(low>high) return -1;
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(mid==nums.size() -1 || nums[mid+1]>target){
                    return mid;
                } else return findlast(nums,mid+1,high,target);
            } 
            else if(target > nums[mid]){
                return findlast(nums,mid+1,high,target);
            }
            return findlast(nums,low,mid-1,target);
        }
};

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def findfirst(self,nums: List[int], t: int) -> int:
        n = len(nums)
        l=0;h=n-1
        lb = -1
        while l<=h:
            m = (l+h)//2
            if nums[m]>= t:
                lb = m
                h = m-1
            else: l = m+1
        return lb
    def findlast(self,nums: List[int], t: int) -> int:
        n = len(nums)
        l=0;h=n-1
        ub = n
        while l<=h:
            m = (l+h)//2
            if nums[m] > t:
                ub = m
                h = m-1
            else: l = m+1
        return ub-1
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        first = self.findfirst(nums,target)
        if first == -1 or nums[first] != target:
            return [-1,-1]
        last = self.findlast(nums, target)
        return [first,last]
        
    


# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna